swarm/network/discovery_test.go

```diff
@@ -58,3 +68,164 @@ func TestDiscovery(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+const (
+	maxPO     = 10
+	maxPeerPO = 8
+)
+
+// TestInitialPeersMsg tests if peersMsg response to incoming subPeersMsg is correct
+func TestSubPeersMsg(t *testing.T) {
+	for po := 0; po < maxPO; po++ {
+		for depth := 0; depth < maxPO; depth++ {
+			t.Run(fmt.Sprintf("PO=%d,advetised depth=%d", po, depth), func(t *testing.T) {
+				testSubPeersMsg(t, po, depth)
+			})
+		}
+	}
+}
+
+// testSubPeersMsg tests that the correct set of peer info is sent
+// to another peer after receiving their subPeersMsg request
+func testSubPeersMsg(t *testing.T, peerPO, peerDepth int) {
+	// generate random pivot address
+	prvkey, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	pivotAddr := pot.NewAddressFromBytes(PrivateKeyToBzzKey(prvkey))
+	// generate control peers address at peerPO wrt pivot
+	peerAddr := pot.RandomAddressAt(pivotAddr, peerPO)
+	// construct kademlia and hive
+	to := NewKademlia(pivotAddr[:], NewKadParams())
+	hive := NewHive(NewHiveParams(), to, nil)
+
+	// expected addrs in peersMsg response
+	var expBzzAddrs []*BzzAddr
+	addrAt := func(a pot.Address, po int) []byte {
+		b := pot.RandomAddressAt(a, po)
+		return b[:]
+	}
+	connect := func(base pot.Address, po int) *BzzAddr {
+		on := addrAt(base, po)
+		peer := newDiscPeer(on)
+		hive.On(peer)
+		return peer.BzzAddr
+	}
+	register := func(base pot.Address, po int) {
+		hive.Register(&BzzAddr{OAddr: addrAt(base, po)})
+	}
+
+	for po := maxPeerPO; po >= 0; po-- {
+		// create a fake connected peer at po from peerAddr
+		on := connect(peerAddr, po)
```

Comment:
This setup will always have one single connected peer per bin.

I tried a simple "hack" trying to add a second peer per bin:
```
for po := maxPeerPO; po >= 0; po-- {
￼ 		// create a fake connected peer at po from peerAddr
￼ 		on := connect(peerAddr, po)
               on2 := connect(peerAddr, po)  <--
// create a fake registered address at po from peerAddr
￼ 		register(peerAddr, po)
￼ 		// we collect expected peer addresses only up till peerPO
￼ 		if po < peerDepth {
￼ 			continue
￼ 		}
￼ 		expBzzAddrs = append(expBzzAddrs, on)
                expBzzAddrs = append(expBzzAddrs, on2)   <---
￼ 	}
```

and the test fails - not sure if that's a valid configuration/approach though.

Reply:
well spotted

Reply:
guaranteeing ordering on the same PO is a bit of a pain, hence i just kept it to one peer per bin.

Reply:
Sorry I don't understand? What is the problem with the order?
Reality is that we usually have 2 peers per bin in a healthy network, so if we have two peers per bin here, the test fails. Is this not relevant?

Reply:
there is no problem it is just that if you generate several peers in the same PO, their order is not immediately trivial to guess, i did not want to compilate the test. we know the iterator itself works, just need to test that peers in the relevant bins are passed.

Reply:
One way is creating a type `SortedBzzAddrs []*BzzAddr`, have this implement `Sort` interface (sort by `.Over()`), then cast the `peers` in `handleSubPeersMsg` to that type, and sort it, do the same with `expBzzAddrs`. I have tried and it passes.

But probably not worth the effort, as it is only of use for this test